test/jdk/java/net/httpclient/FileChannelPublisherTest.java

```diff
@@ -0,0 +1,693 @@
+
+    private static final ServerRequestPair HTTP2 = ServerRequestPair.of(Version.HTTP_2, false);
+
+    private static final ServerRequestPair HTTPS2 = ServerRequestPair.of(Version.HTTP_2, true);
+
+    private static SSLContext createSslContext() {
+        try {
+            return new SimpleSSLContext().get();
+        } catch (IOException exception) {
+            throw new UncheckedIOException(exception);
+        }
+    }
+
+    private record ServerRequestPair(
+            String serverName,
+            HttpTestServer server,
+            BlockingQueue<byte[]> serverReadRequestBodyBytes,
+            HttpRequest.Builder requestBuilder,
+            boolean secure) {
+
+        private static CountDownLatch SERVER_REQUEST_RECEIVED_SIGNAL = null;
+
+        private static CountDownLatch SERVER_READ_PERMISSION = null;
+
+        private static ServerRequestPair of(Version version, boolean secure) {
+
+            // Create the server
+            SSLContext sslContext = secure ? SSL_CONTEXT : null;
+            HttpTestServer server = createServer(version, sslContext);
+            String serverName = secure ? version.toString().replaceFirst("_", "S_") : version.toString();
+
+            // Add the handler
+            String handlerPath = "/%s/".formatted(CLASS_NAME);
+            BlockingQueue<byte[]> serverReadRequestBodyBytes =
+                    addRequestBodyConsumingServerHandler(serverName, server, handlerPath);
+
+            // Create the request builder
+            String requestUriScheme = secure ? "https" : "http";
+            // `x` suffix in the URI is not a typo, but ensures that *only* the parent handler path is matched
+            URI requestUri = URI.create("%s://%s%sx".formatted(requestUriScheme, server.serverAuthority(), handlerPath));
+            HttpRequest.Builder requestBuilder = HttpRequest.newBuilder(requestUri).version(version);
+
+            // Create the pair
+            ServerRequestPair pair = new ServerRequestPair(serverName, server, serverReadRequestBodyBytes, requestBuilder, secure);
+            pair.server.start();
+            LOGGER.log("Server[%s] is started at `%s`", pair, server.serverAuthority());
+
+            return pair;
+
+        }
+
+        private static HttpTestServer createServer(Version version, SSLContext sslContext) {
+            try {
+                // The default HTTP/1.1 test server processes requests sequentially.
+                // This causes a deadlock for concurrent tests such as `testSlicedUpload()`.
+                // Hence, explicitly providing a multithreaded executor for HTTP/1.1.
+                ExecutorService executor = Version.HTTP_1_1.equals(version) ? EXECUTOR : null;
+                return HttpTestServer.create(version, sslContext, executor);
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        }
+
+        private static BlockingQueue<byte[]> addRequestBodyConsumingServerHandler(
```

Comment:
Nit - instead of the long name, it might be better to rename the method to just `addHandler()` and then add a one-liner comment to the method saying that the handler consumes the entire request body.